docs/digits_diffusion_model.md

```diff
@@ -0,0 +1,822 @@
+- Progressive timestep sampling strategy
+- [Exponential Moving Average (EMA)](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average) loss tracking
+- Adaptive noise generation
+
+```{code-cell}
+---
+colab:
+  base_uri: https://localhost:8080/
+id: ZnQqHCAoVfi1
+outputId: a105e2de-ba88-44d0-bad5-3a9a69e54826
+---
+# Initialize training metrics
+losses: List[float] = []                # Store EMA loss history
+moving_avg_loss: Optional[float] = None  # EMA of the loss value
+beta: float = 0.99                      # EMA decay factor for loss smoothing
+
+for epoch in range(num_epochs + 1):
+    # Split PRNG key for independent random operations
+    key, subkey1 = jax.random.split(key)
+    key, subkey2 = jax.random.split(key)
+
+    # Progressive timestep sampling - weights early steps more heavily as training progresses
+    # This helps model focus on fine details in later epochs while maintaining stability
+    progress = epoch / num_epochs
+    t_weights = jnp.linspace(1.0, 0.1 * (1.0 - progress), num_steps)
+    t = jax.random.choice(
+        subkey1,
+        num_steps,
+        shape=(images_train.shape[0],),
+        p=t_weights/t_weights.sum()
+    )
+
+    # Generate Gaussian noise for current batch
+    noise = jax.random.normal(subkey2, images_train.shape)
+
+    # Execute training step with noise prediction and parameter updates
+    loss = train_step(
+        model, optimizer, images_train, t, noise,
+        diffusion.sqrt_alpha_cumulative, diffusion.sqrt_one_minus_alpha_cumulative
+    )
+
+    # Update exponential moving average of loss for smoother tracking
+    if moving_avg_loss is None:
+        moving_avg_loss = loss
+    else:
+        moving_avg_loss = beta * moving_avg_loss + (1 - beta) * loss
+
+    losses.append(moving_avg_loss)
+
+    # Log training progress at regular intervals
+    if epoch % 100 == 0:
+        print(f"Epoch {epoch}, Loss: {moving_avg_loss:.4f}")
+
+print("\nTraining completed.")
+```
+
++++ {"id": "SHX69WgxKUs7"}
+
+
+
++++ {"id": "s-iqch4HKlBV"}
+
+## Training loss visualization
+
+It uses a logarithmic scale to better display the exponential decay of the loss values over time. This representation helps identify both early rapid improvements and later fine-tuning phases of the training process.
+
+```{code-cell}
+---
+colab:
+  base_uri: https://localhost:8080/
+  height: 487
+id: 1bjvWNCcbN24
+outputId: 457fd13f-377f-4021-ddc2-e36940b42550
+---
+# Plot the training loss history with logarithmic scaling
+plt.figure(figsize=(10, 5))            # Create figure with wide aspect ratio for clarity
+plt.plot(losses)                       # losses: List[float] - historical EMA loss values
+plt.title('Training Loss Over Time')
+plt.xlabel('Epoch')
+plt.ylabel('Loss')
+plt.yscale('log')                      # Use log scale to better visualize exponential decay
+plt.grid(True)                         # Add grid for easier value reading
+plt.show()
+```
+
```

Comment:
Suggestion: add some comment explaining what we can see from this visualization (i.e., "We can see that our model performs well in this dataset, due to the flat curve in the right-most part of the graph.")